### PR TITLE
Remove delete() call from CompressorFileStorage.get_available_name

### LINKED ISSUE #1099: Intermittent failure in offline compression
Hi,

We use `django-compressor` (v3.1) in our Django (v4.0.2) application. Our CI randomly fails (something like 1 run in 20) on the compress step with errors like this one:

```
#18 [16/17] RUN python manage.py compress --extension .html,.xsl
#18 3.681 CommandError: An error occurred during rendering registration/password_change_form.html: [Errno 2] No such file or directory: '/code/sereema/staticfiles/CACHE/js/output.5840362cca8f.js'
```

If I read correctly, this is a known issue (#1082) that was fixed in v3.0. But it seems like the fix was not enough.

### PR DESCRIPTION
**Pull request by cuu508**
This PR is an alternative to #1105 and #1100, and intends to fix #1099. It changes CompressorFileStorage:

* drops the deletion logic in  `CompressorFileStorage.get_available_name()`
* this means that `get_available_name` will sometimes return filenames with an underscore and a random string appended at the end to ensure uniqueness
* overrides the `save()` method to add one additional last step: if the filename contains an underscore then move it into the correct location. 

So, instead of deleting a file and then writing new data in the same filename, compressor would write the data in a temporary file, then move it to the correct place. Python docs say `os.replace` is an atomic operation, so this would avoid the brief moment of file not existing between delete and write.

Notes:

* I didn't add any new test cases. In theory there could be a test case which creates compressed files with clashing filenames, and then tests if the produced files contain any underscores (they shouldn't). But not sure if it's worth testing that.
* `GzipCompressorFileStorage` and `BrotliCompressorFileStorage` override the `save` method again and do some more file IO. I'm not sure if we could run into race conditions there too.

### CODE REVIEW

File: compressor/storage.py [hunk 1]
```diff
@@ -33,13 +33,12 @@ def created_time(self, name):
...
-    def get_available_name(self, name, max_length=None):
-        """
-        Deletes the given file if it exists.
-        """
-        if self.exists(name):
-            self.delete(name)
-        return name
+    def save(self, filename, content):
+        temp_filename = super().save(filename, content)
+        if "_" in temp_filename:
```
Comment (diox) [inline compressor/storage.py:38]: Why not compare `temp_filename` with `filename` instead ?

Comment (cuu508) [reply compressor/storage.py]: Ah, of course! Thanks – fixed!

**Comment by diox** [comment]
> I didn't add any new test cases. In theory there could be a test case which creates compressed files with clashing filenames, and then tests if the produced files contain any underscores (they shouldn't). But not sure if it's worth testing that.

I think it's worth testing (I know the current `CompressorFileStorage` implementation doesn't have any tests - we should change that). Something along the lines of what you suggest should be enough. Let me know if you don't have the motivation to do that (you already done a lot of work, I appreciate the time spent on this!) and I'll step in if needed.

> `GzipCompressorFileStorage` and `BrotliCompressorFileStorage` override the save method again and do some more file IO. I'm not sure if we could run into race conditions there too.

They do this with the low-level primitives directly though, so they might run into different issues. I think it's fine to ignore them for this.

**Comment by cuu508** [comment]
No problem – added a small testcase which saves the same file and tests if the filename returned by `save()` contains any underscores or not.

**Review by diox — APPROVED** [review]
:+1: let's merge and get some people experiencing issues with this to test it!

**Pull request merged**